document sigpipe handling (#2909)
aws/s2n-tls: docs/USAGE-GUIDE.md

### s2n\_connection\_set\_fd

```c
int s2n_connection_set_fd(struct s2n_connection *conn,
                          int readfd);
int s2n_connection_set_read_fd(struct s2n_connection *conn,
                               int readfd);
int s2n_connection_set_write_fd(struct s2n_connection *conn,
                                int writefd);
```

**s2n_connection_set_fd** sets the file-descriptor for an s2n-tls connection. This

> file-descriptor should be active and connected; one process is reading from a pipe
> before another writes to it. s2n-tls also supports setting the read and write

Reviewer: I worry this isn't clear enough and should just be removed
```suggestion
file-descriptor should be active and connected. s2n-tls also supports setting the read and write
```

file-descriptors to different values (for pipes or other unusual types of I/O).

**Important Note:**

> If the reading process fails to start or is terminated unexpectedly, writing
> to the pipe will raise a SIGPIPE signal. A SIGPIPE signal will cause the process to

Reviewer: This isn't completely accurate, since it's the pipe we care about and not the process.
```suggestion
If the read end of a pipe is closed unexpectedly, writing
to the pipe will raise a SIGPIPE signal. A SIGPIPE signal will cause the process to
```

read and write file-descriptors to different values (for pipes or other unusual
types of I/O).

Important Note:
Applications should be designed so that one process is reading from a pipe before another

> writes to it. If the reading process fails to start or is terminated unexpectedly, writing

Reviewer: This could be considered a repeat of "This file-descriptor should be active and connected" above. Maybe replace your first line with that instead?

I'd also bold / underline / do something to make "Important Note" stand out.

Author: do you mean delete “Applications should be designed..” or combine them?

The “Applications should be designed..” came from gnu documentation and it seemed more clear even for those not familiar with C development. I would prefer to combine or remove "This
file-descriptor should be active"

direction in which s2n-tls is blocked.

s2n-tls I/O functions should be called repeatedly until the **blocked** parameter is
**S2N_NOT_BLOCKED**.

Applications should be designed so that one process is reading from a pipe before another
writes to it. If the reading process fails to start or is terminated unexpectedly, writing
to the pipe will raise a SIGPIPE signal. A SIGPIPE signal will cause the process to
terminate unless it is handled or ignored. **s2n-tls does NOT handle SIGPIPE.** It is the

> responsibility of the application using s2n-tls to either handle or ignore SIGPIPE.

Reviewer: I think we can say this shorter and clearer, like above. We shouldn't have to go super deep into how pipes work, we just need to remind people that signals exist and need to be ignored/handled.
